### PR TITLE
updated solve_backwards to be fully sparse

### PR DESCRIPTION
**Pull request by amirbarda**
patch for issue #5

### CODE REVIEW

File: torch_sparse_solve.cpp [hunk 1]
```diff
@@ -24,12 +24,16 @@ at::Tensor solve_forward(at::Tensor A, at::Tensor b) {
...
-    at::Tensor gradA_i = A.indices();
-    at::Tensor gradA_v = gradA.index({gradA_i[0], gradA_i[1], gradA_i[2]});
-    at::Tensor gradA_sp = at::sparse_coo_tensor(gradA_i, gradA_v, {at::size(A, 0), at::size(A, 1), at::size(A, 2)});
-    return {gradA_sp, gradb};
+    at::Tensor gradb = solve_forward(A.transpose(1,2).coalesce(), grad);
+    int batch_size = A.size(0);
+    std::vector<at::Tensor>  gradA_list;
+    for (int i=0; i<batch_size; i++) {
+        at::Tensor indices = A[i].coalesce()._indices();
+        at::Tensor gradA_tmp = at::sparse_coo_tensor(indices, -(gradb.index({i, indices.index({0}), 0}) * x.index({i, indices.index({1}), 0}))).unsqueeze(0);
```
Comment (flaport) [inline torch_sparse_solve.cpp:32]: I think you should replace this line with the following:

```python
        at::Tensor gradA_tmp = at::sparse_coo_tensor(indices, -at::sum((gradb.index({i, indices.index({0})}) * x.index({i, indices.index({1})})), -1)).unsqueeze(0);
```

**Comment by flaport** [comment]
Thanks @amir90 !, This looks good!

However, currently the tests are failing... Maybe consider my comment which fixes this :)

**Comment by amir90** [comment]
Yes, you are correct, I didn't need the sum due to passing a coalesced
sparse matrix to my solver.

בתאריך יום ג׳, 3 באוג׳ 2021, 14:29, מאת Floris Laporte ‏<
***@***.***>:

**Comment by flaport** [comment]
merged. Thanks for your patch, @amir90 !

**Pull request merged**